### PR TITLE
Purchases

### PR DESCRIPTION
**Pull request by bonflintstone**
Well this should be done. What I did:

 - Users can select coffees
 - Can specify items (grind, filter, quantify, size)
 - Can add them to the cart
 - Can come back and edit them
 - Can add more items to the cart
 - Can select them as a subscription in the cart
 - Carts lead to checkout
 - All purchases + subscriptions are listed nicely
 - Prices are calculated
 - Users can purchase
 - Orders and Subscriptions are added to the user
 - The order is dispalyed. This includes all items the user bought exactly once.

Yay!

### CODE REVIEW

**Comment by nghildebrandt** [comment]
Nicee! Only thing that you changed the link for the subscribe now image in the welcome/index. Although, you did it in a nicer way (I used a partial), I also  changed it and I would say that it has more to do with user management than with purchases. So, we'll have a small merge conflict, but im sure we can deal with that :)

**Comment by bonflintstone** [comment]
@nghildebrandt So everything is fine then? Why dont ya hit the merge button then
Btw ist there any magic you dont understand? If so comment on it before you approve and I will explain

File: app/controllers/carts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class CartsController < ApplicationController
+  def show
+    @cart = Cart.current(current_user)
```
Comment (nghildebrandt) [inline app/controllers/carts_controller.rb]: Cart.current -> is that Ruby to get the current instance of an object?

File: app/controllers/items_controller.rb [hunk 2]
```diff
@@ -1,10 +1,44 @@
...
+  def destroy
+    Item.find(params[:id]).destroy
+
+    redirect_to cart_path
+  end
+
+  private
+
+  def item_params
+    params.require(:item).permit(:grind, :roast, :coffee_pricing_id, :quantity,
```
Comment (nghildebrandt) [inline app/controllers/items_controller.rb]: still need to look into polymorphism a bit. Gonna read up on it and on Saturday you can explain it to me

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  def new
+    @cart = Cart.current(current_user)
+
+    @cart.remove_empty_subscriptions!
+  end
+
+  def create
+    @cart = Cart.current(current_user)
+
+    order = current_user.orders.create(items: @cart.all_items.map(&:clone))
```
Comment (nghildebrandt) [inline app/controllers/orders_controller.rb]: How does this work exactly with the method clone from item.rb. Is it for adding several items?

File: app/models/subscription.rb [hunk 4]
```diff
@@ -1,5 +1,17 @@
 class Subscription < ApplicationRecord
-  has_many :items, as: :item_holdable
+  has_many :items, as: :item_holdable, dependent: :destroy
 
   belongs_to :subscription_holdable, polymorphic: true
+
+  validates :period_in_weeks, uniqueness: { scope: :subscription_holdable }
+
+  PERIOD_OPTIONS = [1, 2, 4]
```
Comment (nghildebrandt) [inline app/models/subscription.rb]: Some people also like periods that are not in the power of 2 :)

File: app/controllers/carts_controller.rb (hunk 1 above)
Comment (bonflintstone) [reply app/controllers/carts_controller.rb]: No, I wrote that mnethod. The notion of a 'current' object does not really exist outside of this (after all, what would that be?). The helper method `current_user` comes from devise, and I thought I copy the naming to look for a cart by a user.

We could also have used Cart.where(user: current_user).first I suppose

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (bonflintstone) [reply app/controllers/orders_controller.rb]: So basically an item can only ever have one item holdable (after all an item only is one row having one `item_holdable_id` and `..._type`), but when you purchase an item in a subscription you want to have it both saved in the order (so we know what the user ordered) as well as saved in the subscription for the user (so we know what the user subscribed to).

What this line of code exactly does is: create a new order for the current user and copy (new db rows) all the items in the cart into that. The clone method I also implemented myself.

File: app/models/subscription.rb (hunk 4 above)
Comment (bonflintstone) [reply app/models/subscription.rb]: Great thing about this is we can always ask stean and add whatever periods he likes. Localized constants and all, this is just config.

**Pull request merged**